cunumeric/array.py

```diff
@@ -1252,6 +1252,92 @@ def dump(self, file):
+        elif kind == FFTCode.FFT_Z2D:
+            fft_output_type = np.float64
+        elif kind == FFTCode.FFT_C2R:
+            fft_output_type = np.float32
+
+        # Axes and sizes
+        user_axes  = axes is not None
+        user_sizes = s is not None
+        if not user_axes:
+            fft_axes = list(range(-len(s),0)) if user_sizes else list(range(self.ndim))
+        else:
+            fft_axes = axes
+        fft_axes = [x % self.ndim for x in fft_axes]
+
+        fft_s = list(self.shape)
+        if user_sizes:
+            for idx, ax in enumerate(fft_axes):
+                fft_s[ax] = s[idx]
+
+        # Shape
+        fft_input        = self
+        fft_output_shape = self.shape
+        if user_sizes:
+            # Zero padding if any of the user sizes is larger than input
+            zero_padded_input = self
+            if np.any(np.greater(fft_s, fft_input.shape)):
+                # Create array with superset shape, fill with zeros, and copy input in
+                max_size = tuple(np.maximum(fft_s,fft_input.shape))
+                zero_padded_input = ndarray(shape=max_size, dtype=fft_input.dtype)
+                zero_padded_input.fill(0)
+                zero_padded_input._thunk.set_item(tuple(slice(0,i) for i in fft_input.shape), fft_input._thunk)
+
+            # Slicing according to final shape
+            fft_input_shape = list(fft_input.shape)
+            for idx, ax in enumerate(fft_axes):
+                fft_input_shape[ax] = s[idx]
+            fft_input_shape = tuple(fft_input_shape)
+            fft_input  = ndarray(shape=fft_input_shape, thunk=zero_padded_input._thunk.get_item(tuple(slice(0,i) for i in fft_s))).copy()
+            fft_output_shape = fft_input_shape
+
+        # R2C/C2R require different output shapes
+        if fft_output_type != self.dtype:
+            fft_output_shape = list(fft_output_shape)
+            # R2C/C2R dimension is the last axis
+            if direction == FFTDirection.FORWARD:
+                fft_output_shape[fft_axes[-1]] = fft_output_shape[fft_axes[-1]]//2 + 1
+            else:
+                fft_output_shape[fft_axes[-1]] = s[-1] if user_sizes else 2 * (fft_input.shape[fft_axes[-1]] - 1)
+            fft_output_shape = tuple(fft_output_shape)
+
+        # Execute FFT backend
+        out = ndarray(
+            shape=fft_output_shape,
+            dtype=fft_output_type,
+        )
+        fft_input._thunk.fft(out._thunk, fft_axes, kind, direction)
+
+        # Normalization
+        fft_normalization = FFTNormalization.from_string(norm)
+        do_normalization  = (fft_normalization == FFTNormalization.ORTHOGONAL)
+        do_normalization  = do_normalization or (fft_normalization == FFTNormalization.FORWARD and direction == FFTDirection.FORWARD)
+        do_normalization  = do_normalization or (fft_normalization == FFTNormalization.INVERSE and direction == FFTDirection.INVERSE)
+        if do_normalization:
```

Comment:
Maybe something like
```
prod_dims = np.product(norm_shape_along_axes)
if normalize_full
    return out / prod_dims
elif normalize_orthogonal
    return out / sqrt(prod_dims)
else
    return out
```
would be a little bit more readable?

Reply:
I changed it to use `any()`, I hope it will be more readable now :smile: